README
RNAz  1.0
============

Content
=======

0. Terms of use
1. Introduction
2. Theory
   2.1. The structure conservation index
   2.2. Thermodynamic stability
   2.3. Classification based on both scores
3. Usage
   3.1. Installation
   3.2. Enviroment variable
   3.3. Invocation
   3.4. Output
4. Citing RNAz
5. Important notes
6. Contact

0. Terms of use
===============

Please read the file COPYING for licence terms of RNAz. 

1. Introduction
===============

RNAz detects stable and conserved RNA secondary structures in multiple
sequence alignments. RNAz calculates two independent scores for structural
conservation (the structure conservation index SCI) and for thermodynamical
stability (the z-score). High structural conservation (high SCI) and
thermodynamical stability (negative z-scores) are typical features of
functional RNAs (e.g. non-coding RNAs or cis-acting regulatory
elements). RNAz uses both scores to classify a given alignment as
functional RNA or not. It uses a support vector machine classification
procedure which estimates a RNA-class probability which can be used as
convenient overall-score.

For a detailed coverage of all aspects of RNAz we recommend to read the
manual/tutorial (manual.pdf).


2. Theory
=========

2.1. The structure conservation index
=====================================

RNAz uses programs from the Vienna RNA package to perform minimum free
energy (MFE) RNA secondary structure predictions.

First, it calculates the average MFEs for all single sequences in the
aligment using RNAfold.

Second the complete alignment is folded using RNAalifold. RNAalifold
implements a consensus folding algorithm which uses essentially the same
algorithms and energy parameters as RNAfold. It calculates a consensus MFE
which is composed of an energy term averaging the energy contributions of
the single sequences and a covariance term rewarding compensatory and
consistent mutations.

If the sequences in the alignment can fold into a common structure, the
average MFE and the consensus MFE will be of similar dimension. If there is
no common structure, the consensus MFE will be higher (i.e. less stable)
than the average MFE of the single sequences.

Based on this intuitive rationale, a structure conservation index is
defined:

                       consensus MFE
                 SCI=  --------------
                        average MFE


The SCI will be around 0 if RNAalifold does not find a consensus structure,
it will be around 1 if the structure is perfectly conserved. A SCI above 1
indicates a perfectly conserved secondary structure which is even supported
by compensatory and/or consistent mutations.

2.2. Thermodynamic stability
============================

The significance of a predicted MFE as calculated by RNAfold is difficult
to interpret in absolute terms. It depends on the length and the base
composition of the sequences (longer sequence => lower MFE, higher
GC-content => lower MFE). Typically the significane of a MFE is estimated
by comparing to many random sequences of the same length and base
composition. If mu is the mean and sigma the standard deviation of the MFEs
of many random sequences a convenient normalized measure for the
significance of the native sequence with MFE m is a z-score:

                          m - mu
                    z = ---------
                          sigma

RNAz can effeciently calculate z-scores without sampling. Negative z-scores
indicate that the native sequence is more stable than the random
background. The unit of z-scores are standard deviations. Random MFEs can
be roughly approximated by a standard normal distribution which gives an
impression of associated P values (e.g. z=-2 => P=0.98).



2.3. Classification based on both scores
========================================

Both scores represent independent diagnostic features of functional
RNAs. RNAz combines both into a overall score using a support vector
machine regression. Depending on the SCI and z, but also the number of
sequences in the alignment and the mean pairwise identity a RNA
class-probability is calculated. It should be noted that the level of the
SCI depends on the sequence similarity. 100% conservation for example
results in a SCI of 1 per definition but do not hold any information for
our purpose. Thus, the support vector machine was taught to interpret the
significance of the SCI depending on the sequence variation.

The confidence level of this RNA-class probability or "RNAz P-value"
slightly varies depending on the properties of the input alignment. In our
tests, P=0.5 and P=0.9 had specificities of 96% and 99%.


3. Usage
========

3.1. Installation
=================

See INSTALL for details.


3.3. Invocation
===============

RNAz [options] [filename]

RNAz takes an alignment file in the ClustalW or MAF format. Available
command line options:

  -h, --help              Print help and exit
  -V, --version           Print version and exit
  -f, --forward           Score forward strand
  -r, --reverse           Score reverse strand
  -b, --both-strands      Score both strands
  -o, --outfile=FILENAME  Output filename
  -w, --window=START-STOP Score window from START to STOP
  -p, --cutoff=FLOAT      Probability cutoff
  -g, --show-gaps         Display alignment with gap  (default=off)
  -s, --predict-strand    Use strand predictor  (default=off)


You can test RNAz on one of the example alignments (installed 
by default in /usr/local/share/RNAz/examples)

cd /usr/local/share/RNAz/examples

RNAz tRNA.aln


3.4. Output
===========

Please refer to the following commented sample output:

Header:

############################  RNAz 1.0 ###############################
 
 Sequences: 4 <- Number of sequences in the alignment
 Columns: 74  <- Columns of the alignments
 Reading direction: reverse <- Strand considered for calculation 
                    ("forward" or "reverse")
 Mean pairwise identity:  74.38 <- in percent
 Mean single sequence MFE: -33.10 <- average RNAfold MFE in kcal/Mol
 Consensus MFE: -30.98 <- consensus MFE calculated by RNAalifold
 Energy contribution: -27.72 <- RNAalifold energy part
 Covariance contribution:  -3.25 <- RNAalifold covariance part
 Combinations/Pair:   1.64 <- Number of different base pair combination
                              per predicted consensus base-pair
 Mean z-score:  -2.68 
 Structure conservation index:   0.94
 SVM decision value:   3.87  <- Internal decision value, probably 
                                not of much interest...
 SVM RNA-class probability: 0.999672 <- Probability estimate for the
                                        decision value, "RNAz P-value"
 Prediction: RNA <- "RNA" if P>=0.5 else "OTHER"
 
######################################################################

Secondary structure predictions:

The next lines summarize the secondary structure predictions in the
following format:

>sequence name
SEQUENCE
STRUCTURE (MFE)

The structure is indicated in dot bracket notation: '.' denotes a unpaired
base, while '(' and ')' denote base pairs.

The last structure is the RNAalifold consensus structure with the consensus
MFE (broken down in energy and covariance contribution)

>AF041468.1
GGGGGUAUAGCUCAGUUGGUAGAGCGCUGCCUUUGCACGGCAGAUGUCAGGGGUUCGAGUCCCCUUACCUCCA
(((((((..((((........)))).(((((.......))))).....(((((.......)))))))))))). ( -31.60)
>X54300
GGGGGUAUAGCUUAGUUGGUAGAGCGCUGCUUUUGCAAGGCAGAUGUCAGCGGUUCGAAUCCGCUUACCUCCA
(((((((..((((.((.(((((....)))))...)).)))).......(((((.......)))))))))))). ( -27.90)
>L00194
GGGGCCAUAGCUCAGUUGGUAGAGCGCCUGCUUUGCAAGCAGGUGUCGUCGGUUCGAAUCCGUCUGGCUCCA
(((((((..((((........))))(((((((.....)))))))...(.(((.......))).)))))))). ( -32.50)
>AY017179
GGGCCGGUAGCUCAGCCUGGGAGAGCGUCGGCUUUGCAAGCCGAAGGCCCCGGGUUCGAAUCCCGGCCGGUCCA
((((((((...((((((((((...((.((((((.....))))))..)))))))))).))......)))))))). ( -40.40)
>consensus
GGGGCUAUAGCUCAGU_UGGUAGAGCGCCGCCUUUGCAAGGCAGAUGUCAGCGGUUCGAAUCCCCUUACCUCCA
(((((((..((((.........)))).((((((.....)))))).....(((((.......)))))))))))). (-30.98 = -27.72 +  -3.25)


6. Citing RNAz
==============

If you use RNAz in your work please cite:

Washietl S, Hofacker IL, Stadler PF
Fast and reliable prediction of noncoding RNAs.
Proc Natl Acad Sci U S A. 102(7):2454-9 (2005)


6. Contact
==========

Please feel free to contact the author for bug-reports, feature-requests,
suggestions etc. Of course I also look forward to hearing from you if you
discoverd something biologically interesting using RNAz. 

Stefan Washietl <wash@tbi.univie.ac.at>

--
$Id: README,v 1.5 2006/10/12 16:40:24 wash Exp $



NEWS
Version 1.0
===========

Final release of version 1.0. Fixed a major bug which caused RNAz to
crash on 64 bit machines. Fixed a series of smaller bugs in RNAz and
the helper scripts. Thanks to all people who reported bugs.


Version 1.0pre
==============

- RNAz reads MAF alignments in addition to CLUSTAL W format.

- RNAz reads and processes alignments from the input until an "end of file"
  signal is reached. In other words, you can feed as many alignments to RNAz
  as you want. This speeds up analysis since initialization has only be done
  once.

- A set of Perl programs is provided which automatize the
  pre-processing of alignments and thus make large scale screens
  easier.

- A manual/tutorial is available covering all aspects of using
  RNAz and interpreting the results.

- No environment variable is needed any longer making installation
  easier.

- A Windows installer is available.

- As an experimental feature, an additional SVM was added to guess the
  correct strand of a hit (thanks to Kristin Missal).

- Bugs have been fixed, most notably the strange problem with mangled
  output that occured on some systems. Also the uppercase/lowercase problem
  has been fixed.

- The output has been modified. The classification is now named "RNA" and
  "OTHER" (instead of "no RNA"). The "Combinations/Pair" value was added.
  The output can optionally be displayed with gaps, allowing to recover the
  complete input alignment from the RNAz output.


Version 0.1.1
=============

First publicly available version, so everything is new...


--
$Id: NEWS,v 1.6 2006/10/12 16:40:24 wash Exp $



rnaz/RNAz.c
/*********************************************************************                
 *                                                                   *
 *                              RNAz.c                               *
 *                                                                   *
 *	Assess alignments for exceptionally stable and/or conserved      *
 *	secondary structures using RNAalifold/RNAfold and SVMs.          *
 *                                                                   *
 *	          c Stefan Washietl, Ivo L Hofacker                      *
 *                                                                   *
 *	   $Id: RNAz.c,v 1.13 2008-01-24 10:19:14 wash Exp $             *
 *                                                                   *
 *********************************************************************/

#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <unistd.h>
#include <string.h>
#include "fold.h"
#include "fold_vars.h"
#include "utils.h"
#include "pair_mat.h"
#include "alifold.h"
#include "PS_dot.h"
#include "zscore.h"
#include "rnaz_utils.h"
#include "svm.h"
#include "svm_helper.h"
#include "cmdline.h"
#include "strand.h" 

#define IN_RANGE(LOWER,VALUE,UPPER) ((VALUE <= UPPER) && (VALUE >= LOWER))

PRIVATE void usage(void);
PRIVATE void help(void);
PRIVATE void version(void);
PRIVATE void classify(double* prob, double* decValue,
					  struct svm_model* decision_model,
					  double id,int n_seq, double z,double sci);

PRIVATE void gfxPlot(char* fname, char* structure, char* consensus, struct aln *window[],int n_seq);
PRIVATE char **annote(const char *structure, const char *AS[]);


PRIVATE void warning(char* string, double id,int n_seq, double z,double sci,
					 struct aln *AS[]);

enum {FORWARD=1, REVERSE=2};

/********************************************************************
 *                                                                  *
 * main -- main program                                             *
 *                                                                  *
 ********************************************************************/

int main(int argc, char *argv[])
{

  struct svm_model* decision_model; /* SVM classification model */

  /* Command line options */
  int reverse=0;     /* Scan reverse complement */
  int showVersion=0; /* Shows version and exits */
  int showHelp=0;    /* Show short help and exits */
  int from=-1;       /* Scan slice from-to  */
  int to=-1;
  FILE *clust_file=stdin; /* Input file */
  FILE *out=stdout; /* Output file */

  struct aln *AS[MAX_NUM_NAMES];     
  struct aln *window[MAX_NUM_NAMES]; 
  struct aln *windowFwdMem[MAX_NUM_NAMES];

  char *tmpAln[MAX_NUM_NAMES];

  int n_seq;  /* number of input sequences */
  int length; /* length of alignment/window */

  char *structure=NULL;
  char *singleStruc,*gapStruc, *output,*woGapsSeq;
  char strandString[8];
  char warningString[1000];
  char gfxName[100];
  char *string=NULL;
  double singleMFE,sumMFE,singleZ,sumZ,z,sci,id,decValue,prob,comb;
  double min_en, real_en;
  int i,j,k,l,ll,r,countAln;
  int (*readFunction)(FILE *clust,struct aln *alignedSeqs[]);
  char** lines=NULL;
  int directions[3]={FORWARD,0,0};
  int currDirection;
  struct gengetopt_args_info args;

  double meanMFE_fwd=0;
  double consensusMFE_fwd=0;
  double sci_fwd=0;
  double z_fwd=0;
  double gu=0;
  double gu_fwd=0;

  int strandGuess;
  double strandScore,strandProb,strandDec;
  
  if (cmdline_parser (argc, argv, &args) != 0){
	usage();
	exit(EXIT_FAILURE);
  }

  if (args.help_given){
	help();
	exit(EXIT_SUCCESS);
  }

  if (args.version_given){
	version();
	exit(EXIT_SUCCESS);
  }

  if (args.outfile_given){
	out = fopen(args.outfile_arg, "w");
	if (out == NULL){
	  fprintf(stderr, "ERROR: Can't open output file %s\n", args.outfile_arg);
	  exit(1);
	}
  }

  /* Strand prediction implies both strands scored */
  if (args.predict_strand_flag){
    args.both_strands_flag=1;
  }
  
  
  if (args.forward_flag && !args.reverse_flag){
    directions[0]=FORWARD;
    directions[1]=directions[2]=0;
  }
  if (!args.forward_flag && args.reverse_flag){
    directions[0]=REVERSE;
    directions[1]=directions[2]=0;
  }
  if ((args.forward_flag && args.reverse_flag) || args.both_strands_flag){
    directions[0]=FORWARD;
    directions[1]=REVERSE;
  }

  if (args.window_given){
    if (sscanf(args.window_arg,"%d-%d",&from,&to)!=2){
      nrerror("ERROR: Invalid --window/-w command. "
              "Use it like '--window 100-200'\n");
    }
    
    printf("from:%d,to:%d\n",from,to);
  }

  
  if (args.inputs_num>=1){
    clust_file = fopen(args.inputs[0], "r"); 
    if (clust_file == NULL){
      fprintf(stderr, "ERROR: Can't open input file %s\n", args.inputs[0]);
      exit(1);
    }
  }

 
  /* Global RNA package variables */
  do_backtrack = 1; 
  dangles=2;

  switch(checkFormat(clust_file)){

  case CLUSTAL:
    readFunction=&readClustal;
    break;
  case MAF:
    readFunction=&readMaf;
    break;
  case 0:
    nrerror("ERROR: Unknown alignment file format. Use Clustal W or MAF format.\n");
  }

  decision_model=get_decision_model(NULL);
  regression_svm_init(NULL);
  
  countAln=0;

  while ((n_seq=readFunction(clust_file, AS))!=0){

    if (n_seq ==1){
      nrerror("ERROR: You need at least two sequences in the alignment.\n");
    }
	
    countAln++;
	
    length = (int) strlen(AS[0]->seq);
	
    /* if a slice is specified by the user */
  
    if ((from!=-1 || to!=-1) && (countAln==1)){
	  
      if ((from>=to)||(from<=0)||(to>length)){
        nrerror("ERROR: Invalid window range given.\n");
      }
	  
      sliceAln((const struct aln**)AS, (struct aln **)window, from, to);
      length=to-from+1;
    } else { /* take complete alignment */
      /* window=AS does not work..., deep copy seems not necessary here*/
      from=1;
      to=length;
      sliceAln((const struct aln **)AS, (struct aln **)window, 1, length);
    }

    /* Convert all Us to Ts for RNAalifold. There is a slight
       difference in the results. During training we used alignments
       with Ts, so we use Ts here as well. */
    
    for (i=0;i<n_seq;i++){
      j=0;
      while (window[i]->seq[j]){
        window[i]->seq[j]=toupper(window[i]->seq[j]);
        if (window[i]->seq[j]=='U') window[i]->seq[j]='T';
        ++j;
      }
    }
	
    k=0;
    while ((currDirection=directions[k++])!=0){
	  
      if (currDirection==REVERSE){
        revAln((struct aln **)window);
        strcpy(strandString,"reverse");
      } else {
        strcpy(strandString,"forward");
      }

      structure = (char *) space((unsigned) length+1);

      for (i=0;window[i]!=NULL;i++){
        tmpAln[i]=window[i]->seq;
      }
      tmpAln[i]=NULL;
      
      min_en = alifold(tmpAln, structure);
      
      comb=combPerPair(window,structure);
      
      sumZ=0;
      sumMFE=0;

      output=(char *)space(sizeof(char)*(length+16)*(n_seq+1)*3);
  
      for (i=0;i<n_seq;i++){
        singleStruc = space(strlen(window[i]->seq)+1);
        woGapsSeq = space(strlen(window[i]->seq)+1);
        j=0;
        while (window[i]->seq[j]){
          /* Convert all Ts to Us for RNAfold. There is a difference
             between the results. With U in the function call, we get
             the results as RNAfold gives on the command line. Since
             this variant was also used during training, we use it here
             as well. */
          if (window[i]->seq[j]=='T') window[i]->seq[j]='U';
          if (window[i]->seq[j]!='-'){
            woGapsSeq[strlen(woGapsSeq)]=window[i]->seq[j];
            woGapsSeq[strlen(woGapsSeq)]='\0';
          }
          ++j;
        }
	  
        singleMFE = fold(woGapsSeq, singleStruc);
        singleZ=mfe_zscore(woGapsSeq,singleMFE);
        
        sumZ+=singleZ;
        sumMFE+=singleMFE;

        if (window[1]->strand!='?' && !args.window_given){
          sprintf(output+strlen(output),
                  ">%s %d %d %c %d\n",
                  window[i]->name,window[i]->start,
                  window[i]->length,window[i]->strand,
                  window[i]->fullLength);
        } else {
          sprintf(output+strlen(output),">%s\n",window[i]->name);
        }

        if (args.show_gaps_flag){

          gapStruc= (char *) space(sizeof(char)*(strlen(window[i]->seq)+1));
          
          l=ll=0;

          while (window[i]->seq[l]!='\0'){
            if (window[i]->seq[l]!='-'){
              gapStruc[l]=singleStruc[ll];
              l++;
              ll++;
            } else {
              gapStruc[l]='-';
              l++;
            }
          }
		  		  
          sprintf(output+strlen(output),"%s\n%s ( %6.2f)\n",
                  window[i]->seq,gapStruc,singleMFE);
		  
		  
        } else {
          sprintf(output+strlen(output),"%s\n%s ( %6.2f)\n",
                  woGapsSeq,singleStruc,singleMFE);
        }
        free(woGapsSeq);
        free(singleStruc);
      }

      {
        int i; double s=0;
        extern int eos_debug;
        eos_debug=-1; /* shut off warnings about nonstandard pairs */
        for (i=0; window[i]!=NULL; i++) 
          s += energy_of_struct(window[i]->seq, structure);
        real_en = s/i;
      }

      string = consensusSeq((const struct aln**) window);
      sprintf(output+strlen(output),
              ">consensus\n%s\n%s (%6.2f = %6.2f + %6.2f) \n",
              string, structure, min_en, real_en, min_en-real_en );

      id=meanPairID((const struct aln**)window);
      z=sumZ/n_seq;

      if (sumMFE==0){ 
        /*Set SCI to 0 in the weird case of no structure in single
          sequences*/
        sci=0;
      } else {
        sci=min_en/(sumMFE/n_seq);
      }

	  
      
      decValue=999;
      prob=0;
	
      classify(&prob,&decValue,decision_model,id,n_seq,z,sci);

      if (args.cutoff_given){
        if (prob<args.cutoff_arg){
          continue;
        }
      }

	  
      strcpy(warningString,"");

      warning(warningString,id,n_seq,z,sci,(struct aln **)window);
	  

      fprintf(out,"\n############################  RNAz "PACKAGE_VERSION"  ##############################\n\n");
      fprintf(out," Sequences: %u\n", n_seq);

      if (args.window_given){
        fprintf(out," Slice: %u to %u\n",from,to);
	  }
      fprintf(out," Columns: %u\n",length);
      fprintf(out," Reading direction: %s\n",strandString);
      fprintf(out," Mean pairwise identity: %6.2f\n", id);
      fprintf(out," Mean single sequence MFE: %6.2f\n", sumMFE/n_seq);
      fprintf(out," Consensus MFE: %6.2f\n",min_en);
      fprintf(out," Energy contribution: %6.2f\n",real_en);
      fprintf(out," Covariance contribution: %6.2f\n",min_en-real_en);
      fprintf(out," Combinations/Pair: %6.2f\n",comb);
      fprintf(out," Mean z-score: %6.2f\n",z);
      fprintf(out," Structure conservation index: %6.2f\n",sci);
      fprintf(out," SVM decision value: %6.2f\n",decValue);
      fprintf(out," SVM RNA-class probability: %6f\n",prob);
      if (prob>0.5){
        fprintf(out," Prediction: RNA\n");
      }
      else {
        fprintf(out," Prediction: OTHER\n");
      }
      fprintf(out,"%s",warningString);
	  
      fprintf(out,"\n######################################################################\n\n");
	
      fprintf(out,"%s",output);
	
      fflush(out);

      if (args.plot_given){
        if (countAln==1){
          strcpy(gfxName,".ps");
        } else {
          sprintf(gfxName,"%i.ps",countAln-1);
        }
        gfxPlot(gfxName, structure, string, window, n_seq);
      }
 

      if (currDirection==FORWARD && args.predict_strand_flag){

        sliceAln((const struct aln**)window, (struct aln **)windowFwdMem, 1, length);
        
        meanMFE_fwd=sumMFE/n_seq;
        consensusMFE_fwd=min_en;
        sci_fwd=sci;
        z_fwd=z;
      
        string = (char*)consensusSeq((const struct aln**) window);
        gu_fwd=(double)frequency_gu_pairs(string, structure);
      
      }

      if (currDirection==REVERSE && args.predict_strand_flag){

        string = (char*)consensusSeq((const struct aln**) window);
        gu=(double)frequency_gu_pairs(string, structure);


        // strand(double deltaSCI, double deltaMeanMFE, double deltaConsMFE, double deltaZ, int n_seq, double id, double consFreqGU, 
        //            double *score, double *prob, double *decValue, char *modelDir, int code)

        printf("inhere\n");

        if (strand(sci_fwd-sci, meanMFE_fwd-(sumMFE/n_seq), 
                   consensusMFE_fwd-min_en, z_fwd-z, n_seq, id, gu_fwd+gu,
                   &strandScore, &strandProb, &strandDec, NULL,CODE_SCI_Z_MEANMFE_CONSMFE)){


          /* if (predict_strand(sci_fwd-sci, meanMFE_fwd-(sumMFE/n_seq), */
          /*                          consensusMFE_fwd-min_en, z_fwd-z, n_seq, id,  */
          /*                          &strandGuess, &strandProb, &strandDec, NULL)){ */

       
          if (strandScore>1){
            fprintf(out, "\n# Strand winner: forward (%.6f)\n",strandDec);
          } else {
            fprintf(out, "\n# Strand winner: reverse (%.6f)\n",strandDec);
          }
        } else {
          fprintf(out, "\n# WARNING: No strand prediction (values out of range)\n");
        }
      }

      free(structure);
      free(output);

    }
    freeAln((struct aln **)AS);
    freeAln((struct aln **)window);
  }

  if (countAln==0){
    nrerror("ERROR: Empty alignment file\n");
  }
  
  svm_destroy_model(decision_model);
  regression_svm_free();
  
  return 0;
}




/********************************************************************
 *                                                                  *
 * classify -- SVM classification depending on various variables    *
 *                                                                  *
 ********************************************************************
 *                                                                  *
 * prob ... pointer where class probability is stored               *
 * decValue ... pointer where decidion value is stored              *
 * svm_model ... pointer to SVMLIB model used for the classificaton *
 * id ... mean pairwise identity of alignment                       *
 * n_seq ... number of sequences in the alignment                   *
 * z ... mean z-score of single sequences in the alignment          *
 * sci ... structure conservation index of alignment                *
 *                                                                  *
 ********************************************************************/


PRIVATE void classify(double* prob, double* decValue,
					  struct svm_model* decision_model,
					  double id,int n_seq, double z,double sci){

  struct svm_node node[5];


  double* value;
  value=(double*)space(sizeof(double)*2);

  node[0].index = 1; node[0].value = z;
  node[1].index = 2; node[1].value = sci;
  node[2].index = 3; node[2].value = id;
  node[3].index = 4; node[3].value = n_seq;
  node[4].index =-1;

  scale_decision_node((struct svm_node*)&node);

  svm_predict_values(decision_model,node,value);
  *decValue=value[0];

  svm_predict_probability(decision_model,node,value);
  *prob=value[0];
  
  
}
/********************************************************************
 *                                                                  *
 * gfxPlot  -- Generates color annotated alignment and RNA          *
 * strucutre as PostScript                                          *  
 *                                                                  *
 ********************************************************************
 *                                                                  *
 * name ... end of filename                                         *
 * structure ... Consensus structure in dot/bracket notation        *
 * consensus ... Consensus sequence                                 *
 * window ... alignment structure                                   *
 * n_seq  ... number of sequences in the alignment                  *
 *                                                                  *
 ********************************************************************/

void gfxPlot(char* name, char* structure, char* consensus, struct aln *window[],int n_seq){ 

  char **A;
  char fname[128];
  int k;
  char **seqs;
  char **names;

  seqs=(char**)space(sizeof(char*)*(n_seq+1));
  names=(char **)space(sizeof(char*)*(n_seq+1));

  for (k=0;k<n_seq;k++){
    seqs[k]=window[k]->seq;
    names[k]=window[k]->name;
  }

  sprintf(fname,"rna%s",name); 

  A = annote(structure, (const char**) seqs);

  (void) PS_rna_plot_a(consensus, structure, fname, A[0], A[1]);

  sprintf(fname,"aln%s",name);

  PS_color_aln(structure, fname,(const char**) seqs,(const char**) names);

  free(A[0]);
  free(A[1]);
  free(seqs);
  free(names);
  
}


PRIVATE char **annote(const char *structure, const char *AS[]) {
  char *ps, *colorps, **A;
  int i, n, s, pairings, maxl;
  short *ptable;
  char * colorMatrix[6][3] = {
    {"0.0 1", "0.0 0.6",  "0.0 0.2"},  /* red    */
    {"0.16 1","0.16 0.6", "0.16 0.2"}, /* ochre  */
    {"0.32 1","0.32 0.6", "0.32 0.2"}, /* turquoise */
    {"0.48 1","0.48 0.6", "0.48 0.2"}, /* green  */
    {"0.65 1","0.65 0.6", "0.65 0.2"}, /* blue   */
    {"0.81 1","0.81 0.6", "0.81 0.2"} /* violet */
  };

  make_pair_matrix();
  n = strlen(AS[0]);
  maxl = 1024;

  A = (char **) space(sizeof(char *)*2);
  ps = (char *) space(maxl);
  colorps = (char *) space(maxl);
  ptable = make_pair_table(structure);
  for (i=1; i<=n; i++) {
    char pps[64], ci='\0', cj='\0';
    int j, type, pfreq[8] = {0,0,0,0,0,0,0,0}, vi=0, vj=0;
    if ((j=ptable[i])<i) continue;
    for (s=0; AS[s]!=NULL; s++) {
      type = pair[encode_char(AS[s][i-1])][encode_char(AS[s][j-1])];
      pfreq[type]++;
      if (type) {
	if (AS[s][i-1] != ci) { ci = AS[s][i-1]; vi++;}
	if (AS[s][j-1] != cj) { cj = AS[s][j-1]; vj++;}
      }
    }
    for (pairings=0,s=1; s<=7; s++) {
      if (pfreq[s]) pairings++;
    }

    if ((maxl - strlen(ps) < 192) || ((maxl - strlen(colorps)) < 64)) {
      maxl *= 2;
      ps = realloc(ps, maxl);
      colorps = realloc(colorps, maxl);
      if ((ps==NULL) || (colorps == NULL))
	  nrerror("out of memory in realloc");
    }

    if (pfreq[0]<=2) {
      snprintf(pps, 64, "%d %d %s colorpair\n",
	       i,j, colorMatrix[pairings-1][pfreq[0]]);
      strcat(colorps, pps);
    }

    if (pfreq[0]>0) {
      snprintf(pps, 64, "%d %d %d gmark\n", i, j, pfreq[0]);
      strcat(ps, pps);
    }
    if (vi>1) {
      snprintf(pps, 64, "%d cmark\n", i);
      strcat(ps, pps);
    }
    if (vj>1) {
      snprintf(pps, 64, "%d cmark\n", j);
      strcat(ps, pps);
    }
  }
  free(ptable);
  A[0]=colorps;
  A[1]=ps;
  return A;
}




/*Hardcoded limits a more sophisticated data-model for meta-model information should be
  considered*/

PRIVATE void warning(char* string, double id,int n_seq, double z,double sci,
					 struct aln *AS[]){

  double GC,A,C;
  int i,j,length,n_A,n_C,n_T,n_G;
  char *seq;

  if (id>100.0) {
	strcpy(string," WARNING: Mean pairwise identity too large.\n");
	string+=strlen(string);
  }

  if (id<52.35) {
	strcpy(string," WARNING: Mean pairwise identity too low.\n");
	string+=strlen(string);
  }

  if (n_seq<2) {
	strcpy(string," WARNING: Too few sequences in alignment.\n");
	string+=strlen(string);
  }
  
  if (n_seq>6) {
	strcpy(string," WARNING: Too many sequences in alignment.\n");
	string+=strlen(string);
  }

  if (!IN_RANGE(-7.87,z,2.76)){
	strcpy(string," WARNING: Mean z-score out of range.\n");
	string+=strlen(string);
  }
  
  if (!IN_RANGE(0,sci,1.23)){
	strcpy(string," WARNING: Structure conservation index out of range.\n");
	string+=strlen(string);
  }

  for (i=0;AS[i]!=NULL;i++){
	seq=AS[i]->seq;

	//printf("SEQ: %s\n",seq);
	
	n_A=n_T=n_G=n_C=0;
  
	for (j=0; j<strlen(seq); j++) {
	  switch (seq[j]) {
	  case 'A': n_A++; break;
	  case 'C': n_C++; break;
	  case 'G': n_G++; break;
	  case 'T': n_T++; break;
	  case 'U': n_T++; break;  
	  }
	}
	
	length=strlen(seq);


	GC=((double)(n_G+n_C)/(double)(n_G+n_C+n_A+n_T));
	A=((double)n_A/(n_A+n_T));
	C=((double)n_C/(n_G+n_C));

	if (length<50){
	  sprintf(string," WARNING: Sequence %d too short.\n",i+1);
	  string+=strlen(string);
	}

	if (length>400){
	  sprintf(string," WARNING: Sequence %d too long.\n",i+1);
	  string+=strlen(string);
	}
	
	if ((!IN_RANGE(0.25,GC,0.75)) ||
		(!IN_RANGE(0.25,GC,0.75)) ||
		(!IN_RANGE(0.25,GC,0.75))){
	  sprintf(string," WARNING: Sequence %d: Base composition out of range.\n",i+1);
	  string+=strlen(string);
	}
  }
}



/********************************************************************
 *                                                                  *
 * usage, help, version - shows information and exits               *
 *                                                                  *
 ********************************************************************/

PRIVATE void usage(void){

  help();

}

PRIVATE void help(void){

  cmdline_parser_print_version ();

  printf("\nUsage: %s [OPTIONS]... [FILES]\n\n", CMDLINE_PARSER_PACKAGE);
  printf("%s\n","  -h, --help              Print help and exit");
  printf("%s\n","  -V, --version           Print version and exit");
  printf("%s\n","  -f, --forward           Score forward strand");
  printf("%s\n","  -r, --reverse           Score reverse strand");
  printf("%s\n","  -b, --both-strands      Score both strands");
  printf("%s\n","  -o, --outfile=FILENAME  Output filename");
  printf("%s\n","  -w, --window=START-STOP Score window from START to STOP");
  printf("%s\n","  -p, --cutoff=FLOAT      Probability cutoff");
  printf("%s\n","  -g, --show-gaps         Display alignment with gap  (default=off)");
  printf("%s\n","  -s, --predict-strand    Use strand predictor  (default=off)");
  printf("%s\n\n","  -x, --plot              Generate graphical output  (default=off)");

}

PRIVATE void version(void){
  printf("RNAz version " PACKAGE_VERSION "\n");
  exit(EXIT_SUCCESS);
}







